### PR TITLE
Convert string args to d_import_dirs.

### PR DESCRIPTION
**Pull request by amcn**
Also update the documentation to reflect the accepted types.

Fixes: #12742

### CODE REVIEW

**Comment by amcn** [comment]
This PR represents the smallest diff that fixes the issue raised in #12742. 

An alternative, and perhaps better, approach would be to fully align the treatment of `d_import_dirs` between `declare_dependancy` and `build_target` (and all its children).

In `declare_dependancy`, `d_import_dirs` is handled this way:

```python
d_import_dirs = self.extract_incdirs(kwargs, 'd_import_dirs')
```
It's worth noting that much of the same logic that is implemented in `kwarg_strings_to_includedirs` is already done by `extract_incdirs`.

**Comment by amcn** [comment]
I mentioned in a previous comment that this PR represents the smallest diff that fixes #12742 but that perhaps a more comprehensive rework would be better (if more consequential).

I have since pushed up a new commit which implements that more consequential version. This version changes meson's behavior in a user visible manner and the tests had to be updated to resolve this.

I'll let maintainers decide which approach is preferable.

File: mesonbuild/interpreter/interpreter.py
Comment (dcbaker) [inline mesonbuild/interpreter/interpreter.py:3488]: This deprecation warning (or should be `mlog.deprecated` instead of `mlog.warning`) is getting lost. It's important, and needs to be ported to extract_incdirs. And the test that ensures this keeps working should not drop this.

**Review by dcbaker — CHANGES_REQUESTED** [review]

File: mesonbuild/interpreter/interpreter.py
Comment (dcbaker) [reply mesonbuild/interpreter/interpreter.py:3488]: Since we already handle some feature changes in extract_incdirs, we should handle this one inside it too, but preferably using a `FeatureDeprecated` mirror the using of the `FeatureNew` inside `extract_incdirs`

Comment (amcn) [reply mesonbuild/interpreter/interpreter.py:3488]: This is now implemented. Let me know what you think.

File: test cases/d/9 features/meson.build
Comment (dcbaker) [inline test cases/d/9 features/meson.build:1]:
```suggestion
project('D Features', 'd', meson_version: '>=1.6', default_options : ['debug=false'])
```
The project test runner has some magic code to handle `.99` versions gracefully. (using either `./run_single_test.py` or `./run_project_tests.py`)

File: mesonbuild/interpreter/interpreter.py
Comment (dcbaker) [inline mesonbuild/interpreter/interpreter.py:2806]:
```suggestion
                                                 '0.45', self.subproject,
```
Did a little git archeology, this is an old deprecation!

File: mesonbuild/interpreter/interpreter.py
Comment (dcbaker) [inline mesonbuild/interpreter/interpreter.py:2807]: We don't need the hard-error bit, deprecation implies that.

```suggestion
                                                 'Use a relative path instead.',
```

**Review by dcbaker — CHANGES_REQUESTED** [review]
A couple of small things, but this is looking really good, thanks!

File: test cases/d/9 features/meson.build
Comment (amcn) [reply test cases/d/9 features/meson.build]: Ah cool, I didn't know that. Fix applied in the latest version.

File: mesonbuild/interpreter/interpreter.py
Comment (amcn) [reply mesonbuild/interpreter/interpreter.py]: Done, many thanks for the archaeological research! This could've caused a lot of confusion for users had you not pointed this correction out.

File: mesonbuild/interpreter/interpreter.py
Comment (amcn) [reply mesonbuild/interpreter/interpreter.py]: Done. I had used the old error message verbatim but I admit that I found it a little bit user hostile. Your suggestion is much better.

**Review by dcbaker — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
